### PR TITLE
I18n generator && OpenEMR 2x faster && new js translation functionllity

### PR DESCRIPTION
**Pull request by matrix-amiel**
Hi!!

Here I suggest a simple mechanism that generates I18n file with all the translations from the database.
The generator runs in the 'npm run build' and creates JSON file for each language.

With the i18n jsons we can implement 2 important changes -
1.  Amazing performance improvement of all the system. Instead of going to the database in every translation, all the translations are fetched from cache (from json files).
This change improves performance 2x!!
For example -
demographics screen
0.28318619728088 (run time with translation from cache)
0.454185962677 (from database)
massages screen
0.090673923492432 (run time with translation from cache)
0.15632295608521 (from database)
globals screen
0.16775393486023 (run time with translation from cache)
0.29873204231262 (from database)

2. Open the option to translation into js files, required for future developing with JS framework (like ReactJS, etc.)  and very usefully for clean writing with separating of js scripts from php files.

How it's working:

- The generator working in the 'npm run build' process or  run itself with 'npm run translation' and creates json file in the `interface/language/i18n` for each langauge.

- In the server side, if i18n file for selected languages exists, all the translation functions take the translate from the json.

- For the client side, in the login page the json of selected language saved in the sessionStorage and available for each page. I built basic file for init i18next library with the translation, and it can be loaded using `Header::setupHeader(['js-translation'])` 
After loading those scripts, possible to translate in the js file -
```
i18n.then(function(t) {
    document.getElementById('output').innerHTML = i18next.t('key');
  });
``` 

I will be happy for comments and thoughts...

Amiel - Matrix

### CODE REVIEW

**Comment by mdsupport** [comment]
@matrix-amiel,
Your db benchmark was surprising since db vendors have spent millions of hours optimizing key=>value retrieval.  By exporting stuff to files, you are creating snapshot(s).  If project is ok with that setup, with its constraints, then better approach will be : 
1. Create de-normalized table that have lang_id, en_text, lang_text as the columns and all columns as PK.
2. logon.php after successful login should run query - select lang_id, count(en_text) where lang_id=?
This query will pull all en_text => lang_text index table pages into mysql buffer which would be shared between all users until the buffer gets exhausted or times out.

Performing translations in js is going to mean touching every script and enclosing the translatable texts into a html tag with class.  If that approach is chosen, you would create a new challenge for i18N dependent installations that, like us, run standard scripts in background cron jobs for automated workflows.

However, if you make this feature as opt-in and it works for your situation, it is a good option to have for others.  Only suggestion would be to move generated files out to documents or some site specific directory.  Also, if you were to include language variable in file name, the js file should become part of browser cache.

**Comment by matrix-amiel** [comment]
Hi @mdsupport 
1. Your suggestion with mysql buffer is interesting, but it's another solution than i18n and I don't know if it improves the performance more than searching in php array (maybe we should add something like memcached instead of load it from the file every runtime..).    
2. why we need cron for i18n files, why update with 'npm run build' not enough?

**Comment by mdsupport** [comment]
@matrix-amiel, suggestion re db-centric processes was more from admin and maintenance perspective than performance. While it may be possible that php is better at use of indexes for accessing data than mysql or other dbs, I would imagine db will be close 2nd.  The real gain in speed, as always, is coming at the expense of concurrency. 

More important point about using JS frameworks relates to php scripts in cli mode or other data centric APIs do not have js endpoint.  Again, just worth 2 cents or less.

**Comment by matrix-amiel** [comment]
Hi, 
In this PR there 2 parts, changes in the translations in server-side and client-side.
Regrading the server side we can discuss how the best performance, and probably it depends on the machine if the DB is strong it can be better.

But I would happy to know what do you think about the solution of JS translation. It required for every JS development we will do.

**Comment by sjpadgett** [comment]
Hi @matrix-amiel Agree we need this type of functionality.
Also may be helpful if you'd include an example use case with this PR. Maybe pick a simple script somewhere.

**Comment by matrix-amiel** [comment]
- Install/switch to my branch.

- Run  'npm run build'  or  'npm run translation', it will create json file in the `interface/language/i18n` for each langauge with all the translation from the DB.

- In the login process the json of selected language will be saved in the sessionStorage and will be available for each page. 

For loading of the translation from sessionStorage to the pages, I built basic file with init i18next library  and it can be loaded using `Header::setupHeader(['js-translation'])` 
After loading this script, possible to translate in the js file -

for example - 
```
i18n.then(function(t) {
    document.getElementById('output').innerHTML = i18next.t('key');
  });
```

**Comment by sjpadgett** [comment]
Fair enough, thanks. 
Question:
How will the translations for use in JS get translated to the database eg say I want a translation 'Hello' for a new dialog?

Doesn't session storage have a size limit? Any chance of bumping up against?

It'd be nice if the translations could be fetched in a group fashion by context eg fetch all translations for context name Alert-One.

**Comment by matrix-amiel** [comment]
The json of the translations is export of all the constants and definitions to files for each language, so the way of adding translation is not changed. 

Fetch translation for each JS page is could be excellent but it requires to create a new logic of saved translations and to maintaining 2 mechanisms at the same time.

**Comment by sjpadgett** [comment]
Maybe i'm missing something. The main reason to have this utility is for JS scripts that stand on their own ie .js. If I want to translate a new constant in the script, how does that get to the database? eq in php, developer does translation, eventually picked up by the translation engine, defined and stored to database.(I'm unsure how that works on backend so I may need to be educated).

Also concerning browser sessionStorage, What are the chances you could run up against a 5mb storage limit? Couldn't you just use memory?

**Comment by matrix-amiel** [comment]
This script creates and updates the JSON files every process of 'npm run build'.
```
while ($lang = SqlFetchArray($languages)) {
    if (!is_file(dirname(__FILE__)) . '/i18n/' . $lang['lang_code'] . '.json') {
        touch(dirname(__FILE__) . '/i18n/' . $lang['lang_code'] . '.json');
    }
    $sql = "SELECT c.constant_name, d.definition FROM lang_definitions as d  
            JOIN lang_constants AS c ON d.cons_id = c.cons_id 
            WHERE d.lang_id = ?";
    $tarns = sqlStatement($sql, $lang['lang_id']);
    $json = array();
    while ($row = SqlFetchArray($tarns)) {
        $json[$row['constant_name']] = $row['definition'];
    }
    file_put_contents(dirname(__FILE__) . '/i18n/' . $lang['lang_code'] . '.json', json_encode($json, JSON_UNESCAPED_UNICODE));
}

``` 
And in the login the JSON is saved in the local session, and available in every js file. 
My idea is to save single storage to all the translation, and do not create new storage for JS files.

Regarding the memory, the largest json file (right now..) is 0.5 MG. so I don't think there is problem os size.

**Comment by sjpadgett** [comment]
hmm i'd thought the json file would be much larger. I will need to go ahead and clone and play with but this all seems reasonable to me.
I'm still not seeing that if I request a translation in JS that is not already in the database, does that new translation get picked up somehow and put in the database. 
@bradymiller How does that work now with xlate in PHP. Do you run a task that scans for new translations to add to database and does that include JS scripts?

**Comment by bradymiller** [comment]
Hi,

@sjpadgett , Here's doc going over the translation pipeline: https://github.com/openemr/openemr/blob/master/contrib/util/language_translations/README
(check out details on the collectConstants.pl script there)

Will be much better to call the json build within OpenEMR (ie. no need to connect it to npm stuff) in strategic places (Checkout Admin->Other->Languages; for now, just make a link in that screen for json build, however end goal will be that any time language is updated there or synchronized will rebuild the json). And then store the jsons in sites/default(or whatever the site it)/documents/language_translations/i18n/ .

**Comment by sjpadgett** [comment]
Thanks @bradymiller I knew it was an involved process but bottom line it for me, we can add scanning javaScript files, correct? Why not build the json files as part of this process(building database)?
The way I recently approached this can be seen in main.php jsFetchGlobals(scope). What makes @matrix-amiel pattern possibly superior to mine is ensuring the translation constants are available to JS routines before routines are called. Due to JavaScript asynchronous nature, this is an issue. ie ensuring translations have scope.  

Something about storing translations in local storage bothers me however, I can't think of a technical reason why not. So, I leave that to others to decide.

Bottom line for me is, we need this capability and will try to do a test implementation here soon. Maybe where I used jsFetchGlobals in dialog alert. 
I'll leave how translations are done to Amiel and Brady and depending on how my test goes, I say this is okay by me.
Amiel, have you guys used this anywhere yet in your project?

**Comment by bradymiller** [comment]
hi @sjpadgett ,
Main issue is that the official translation set always requires customization by installations (this is why there is a synchronize feature in the Language editor (can import official translation set and then synchronize in the local changes). So, need a realtime, seamless (ie. behind the scenes) approach for also supporting customization in the json translation files.

**Comment by bradymiller** [comment]
Was thinking more about this. How expensive (in terms of speed) would it be to cut out the middlemen (json files)?
And instead:
1. Populate $GLOBALS['i18n'] in interface/globals from the database directly
2. In the call to loadI18nToSessionStorage in login.php, this basically calls a modification of the i18n_generator.php script, which returns the json of the requested language (ie. so no need to build/store jsons).

File: interface/login/login.php [hunk 1]
```diff
@@ -393,5 +393,28 @@ function imsubmitted() {
...
+        //The translations will be available for js in each page
+        //For develop js with translation recommended to load the i18next library with Header::setupHeader(['js-translation'])
+        function loadI18nToSessionStorage(lang_code) {
+            $('[type="submit"]').attr('disabled', true);
+            $.ajax( {
+                url: <?php echo js_escape($GLOBALS['webroot'])?>+"/interface/language/i18n/"+lang_code+".json",
+                dataType: 'text'
+            })
+            .done(function (data) {
+                sessionStorage.setItem('i18n', data);
```
Comment (bradymiller) [inline interface/login/login.php]: I think the key for the sessionStorage will need to have the language name and the "site" directory (usually default; ie. the multisite stuff) in it to avoid weird things happening if the same browser has multiple logins in it (using different languages and/or site).
Also, would it make more sense to run this function when user tries to login rather than every time they change the language in the selector?

**Comment by sjpadgett** [comment]
@bradymiller I've patched this PR in to current master for testing. I'll look at your concerns and get back to you guys. I'm curious how this goes with replacing my dialog alert translations. I have high hopes:)

**Comment by sjpadgett** [comment]
As far as calling in JS, this new method needs to be called as I call in my solution. Meaning, we still need to resolve a promise before we implement where xlate occurs in script. Of course, this is pretty much the way it has to happen.

Still, i'd really like to see a solution where we don't need to worry if the translations are in scope to the executing script before we execute routines where the  translations are needed. i.e Make the i18next object available global. Perhaps from main.

This is still a more elegant solution and if we can hammer out(or somebody educates me) how new translations that will occur in javascript scripts, are translated to the database and we clean up how login is handled, we bring this in where we can collaborate further. Brady, i'm looking at your suggestions for login next.
For sessionStorage, I no longer have misgivings. For some reason I thought the translation files would be much larger. Speaking of which, default en language is not created on npm build or I missed something.

If anyone is interested here is how I implemented in dialog for alertMsg routine:
`
// universal alert popup message
if (typeof alertMsg !== "function") {
    function alertMsg(message, timer = 5000, type = 'danger', size = '', persist = '') {
        // example of get i18n xlate.
        i18n.then(t => {
            $('#alert_box').remove();
            let oHidden = '';
            oHidden = !persist ? "hidden" : '';
            let oSize = (size == 'lg') ? 'left:10%;width:80%;' : 'left:25%;width:50%;';
            let style = "position:fixed;top:25%;" + oSize + " bottom:0;z-index:9999;";
            $("body").prepend("<div class='container text-center' id='alert_box' style='" + style + "'></div>");
            let mHtml = '<div id="alertmsg" hidden class="alert alert-' + type + ' alert-dismissable">' +
                '<button type="button" class="btn btn-link ' + oHidden + '" id="dontShowAgain" data-dismiss="alert">' +
                i18next.t("Got It") +'&nbsp;<i class="fa fa-thumbs-up"></i></button>' +
                '<h4 class="alert-heading text-center">' + i18next.t("Alert") + '!</h4><hr>' + '<p style="color:#000;">' + message + '</p>' +
                '<button type="button" class="pull-right btn btn-link" data-dismiss="alert">' + i18next.t("Dismiss") + '</button></br></div>';
            $('#alert_box').append(mHtml);
            $('#alertmsg').fadeIn(800);
            $('#alertmsg').on('closed.bs.alert', function () {
                clearTimeout(AlertMsg);
                $('#alert_box').remove();
                return false;
            });
            $('#dontShowAgain').on('click', function (e) {
                persistUserOption(persist, 1);
            });
            let AlertMsg = setTimeout(function () {
                $('#alertmsg').fadeOut(800, function () {
                    $('#alert_box').remove();
                });
            }, timer);
        }).catch(error => {
            console.log(error.message)
        });
    }
}`

**Comment by bradymiller** [comment]
@sjpadgett , Streamlined the js translation mechanism. Check out this PR: https://github.com/openemr/openemr/pull/2625

**Comment by sjpadgett** [comment]
> I will be happy for comments and thoughts...

Well @matrix-amiel you asked.:)
I thinks Brady's example may be the way to go. Seems to be more streamlined and universal for implementation for developers. You may have a reason for needing the json files that we're not seeing so, please let us know.
Thanks for tackling this much needed mechanism making for one less hurdle for frontend improvements. 
Eagerly awaiting Matrix's thoughts.

**Comment by bradymiller** [comment]
Closing this since just brought in another PR which addresses this: https://github.com/openemr/openemr/pull/2625
